flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateReaderImpl.java

```diff
@@ -0,0 +1,116 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.channel;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.checkpoint.OperatorSubtaskState;
+import org.apache.flink.runtime.checkpoint.TaskStateSnapshot;
+import org.apache.flink.runtime.checkpoint.channel.RefCountingFSDataInputStream.RefCountingFSDataInputStreamFactory;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.BufferBuilder;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.state.AbstractChannelStateHandle;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.guava18.com.google.common.io.Closer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.util.Arrays.asList;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * {@link ChannelStateReader} implementation. Usage considerations:
+ * <ol>
+ *     <li>state of a channel can be read once per instance of this class; once done it returns
+ *     {@link org.apache.flink.runtime.checkpoint.channel.ChannelStateReader.ReadResult#NO_MORE_DATA NO_MORE_DATA}</li>
+ *     <li>reader/writer indices of the passed buffer are respected and updated</li>
+ *     <li>buffers must be prepared (cleared) before passing to reader</li>
+ *     <li>buffers must be released after use</li>
+ * </ol>
+ */
+@NotThreadSafe
+@Internal
+public class ChannelStateReaderImpl implements ChannelStateReader {
+	private static final Logger log = LoggerFactory.getLogger(ChannelStateReaderImpl.class);
+
+	private final Map<InputChannelInfo, ChannelStateStreamReader> inputChannelHandleReaders;
+	private final Map<ResultSubpartitionInfo, ChannelStateStreamReader> resultSubpartitionHandleReaders;
+
+	public ChannelStateReaderImpl(TaskStateSnapshot snapshot) {
+		this(snapshot, new ChannelStateSerializerImpl());
+	}
+
+	@SuppressWarnings("WeakerAccess")
+	ChannelStateReaderImpl(TaskStateSnapshot snapshot, ChannelStateDeserializer serializer) {
+		RefCountingFSDataInputStreamFactory streamFactory = new RefCountingFSDataInputStreamFactory(serializer);
+		final HashMap<InputChannelInfo, ChannelStateStreamReader> inputChannelHandleReadersTmp = new HashMap<>();
```

Comment:
we can pass class field `inputChannelHandleReaders` directly  in below `addReaders` to avoid temporary variables.

Reply:
The consideration here is concurrency: the maps constructed here are read and cleared in `close()` potentially by another thread.
So we need a memory barrier after construction to make the change visible to this thread.
This barrier is a write to `final` variables here.
I'll add a comment about it.

Reply:
I guess you are indicating the concurrent issue. Before the constructor is finished, some other threads might access intermediate `inputChannelHandleReaders` and `resultSubpartitionHandleReaders`. 

If so, it seems not valid because no one can get and reference `ChannelStateReaderImpl` before the constructor finishes. In real codes, the constructing is done before creating `Task` by RPC thread, then no one would reference it to read or close until it is created to pass into `task` class.

Anyway it is not big issue on my side, just to clarify the consideration.

Reply:
No, I meant that  **after** construction, the instance can be accessed by other threads. 
And AFAIK there is no memory barrier after constructor by default.
So without temporary variables, the only guarantee is that `inputChannelHandleReaders` points to a map, but not about contents of that map.

Reply:
From [JLS](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5):
> An object is considered to be completely initialized when its constructor finishes. A thread 
> that can only see a reference to an object after that object has been completely initialized is 
>  guaranteed to see the correctly initialized values for that object's final fields.

I read it as "you have no guarantee about non-final fields". Which means there is no barrier in the end of the constructor.